src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -305,16 +305,15 @@ crate enum FragmentKind {
 
 impl ItemFragment {
     /// Create a fragment for an associated item.
-    ///
-    /// `is_prototype` is whether this associated item is a trait method
-    /// without a default definition.
-    fn from_assoc_item(def_id: DefId, kind: ty::AssocKind, is_prototype: bool) -> Self {
-        match kind {
+    #[instrument(level = "debug")]
+    fn from_assoc_item(item: &ty::AssocItem) -> Self {
+        let def_id = item.def_id;
+        match item.kind {
             ty::AssocKind::Fn => {
-                if is_prototype {
-                    ItemFragment(FragmentKind::TyMethod, def_id)
-                } else {
+                if item.defaultness.has_value() {
                     ItemFragment(FragmentKind::Method, def_id)
+                } else {
+                    ItemFragment(FragmentKind::TyMethod, def_id)
```

Comment:
A method is a method that has a body, while a tymethod is a prototype. The distinction is unnecessary and I already have an open PR that merges them.

What doesn't make sense to you about `FragmentKind`? I'm happy to try to explain it.

Reply:
>What doesn't make sense to you about `FragmentKind`?

The set of variants it has.
Why are associated items (including variants) combined with fields?
What are those text pieces that `fn render` produce and why do they need to be different?

Reply:
A `FragmentKind` represents some sort of "projection" out of a type. So, since fields are conceptually "children" of a type, as are variants and associated items, they get a `FragmentKind`.

The ultimate reason why `FragmentKind` exists is that rustdoc only generates independent pages for some kinds of items. For example, structs and free functions get their own pages, but fields, variants, and associated items are just sections on their parent type's page. In most places, this is captured by returning a `Res` for the page and a `FragmentKind` representing the section within that page. Then the `Res` is turned into a path like `../foo/struct.Bar.html` and the `FragmentKind` is turned into a [URL fragment](https://en.wikipedia.org/wiki/URI_fragment) like `#structfield.baz`.

Does that explanation help?

Reply:
Yes, that makes it more clear.